File: 96. MinimumGeneticMutation.java
import java.util.ArrayDeque;
import java.util.HashSet;
import java.util.Queue;
import java.util.Set;

public class MinimumGeneticMutation {

    public static int minMutation(String startGene, String endGene, String[] bank) {
        Set<String> bankSet = new HashSet<>();
        for (String gene : bank) {
            bankSet.add(gene);
        }

        if (!bankSet.contains(endGene)) {
            return -1;
        }

        char[] choices = {'A', 'C', 'G', 'T'};
        Queue<String> queue = new ArrayDeque<>();
        Set<String> visited = new HashSet<>();

        queue.offer(startGene);
        visited.add(startGene);

        int mutations = 0;

        while (!queue.isEmpty()) {
            int size = queue.size();

            for (int i = 0; i < size; i++) {
                String current = queue.poll();

                if (current.equals(endGene)) {
                    return mutations;
                }

                char[] chars = current.toCharArray();

                for (int j = 0; j < chars.length; j++) {
                    char original = chars[j];

                    for (char choice : choices) {
                        if (choice == original) {
                            continue;
                        }

                        chars[j] = choice;
                        String next = new String(chars);

                        if (bankSet.contains(next) && !visited.contains(next)) {
                            visited.add(next);
                            queue.offer(next);
                        }
                    }

                    chars[j] = original;
                }
            }

            mutations++;
        }

        return -1;
    }

    public static void main(String[] args) {
        System.out.println(minMutation("AACCGGTT", "AACCGGTA", new String[]{"AACCGGTA"}));
        System.out.println(minMutation("AACCGGTT", "AAACGGTA", new String[]{"AACCGGTA", "AACCGCTA", "AAACGGTA"}));
        System.out.println(minMutation("AAAAACCC", "AACCCCCC", new String[]{"AAAACCCC", "AAACCCCC", "AACCCCCC"}));
        System.out.println(minMutation("AACCGGTT", "AACCGGTA", new String[]{}));
    }
}

File: 96. MinimumGeneticMutation.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_BANK 10
#define GENE_LEN 8

static int minMutation(const char *startGene, const char *endGene,
                       const char bank[][GENE_LEN + 1], int bankSize) {
    /* Check if endGene is in bank */
    int endInBank = 0;
    for (int i = 0; i < bankSize; i++) {
        if (strcmp(bank[i], endGene) == 0) { endInBank = 1; break; }
    }
    if (!endInBank) return -1;

    const char choices[] = {'A', 'C', 'G', 'T'};
    int numChoices = 4;

    /* BFS queue */
    char queue[MAX_BANK * 2 + 10][GENE_LEN + 1];
    int  qhead = 0, qtail = 0;
    int  visited[MAX_BANK] = {0};

    strncpy(queue[qtail++], startGene, GENE_LEN + 1);

    int mutations = 0;

    while (qhead < qtail) {
        int size = qtail - qhead;
        for (int i = 0; i < size; i++) {
            char current[GENE_LEN + 1];
            strncpy(current, queue[qhead++], GENE_LEN + 1);

            if (strcmp(current, endGene) == 0) return mutations;

            char next[GENE_LEN + 1];
            for (int j = 0; j < GENE_LEN; j++) {
                char original = current[j];
                for (int k = 0; k < numChoices; k++) {
                    if (choices[k] == original) continue;
                    strncpy(next, current, GENE_LEN + 1);
                    next[j] = choices[k];

                    for (int b = 0; b < bankSize; b++) {
                        if (!visited[b] && strcmp(bank[b], next) == 0) {
                            visited[b] = 1;
                            strncpy(queue[qtail++], next, GENE_LEN + 1);
                        }
                    }
                }
            }
        }
        mutations++;
    }
    return -1;
}

int main(void) {
    {
        const char bank[][GENE_LEN + 1] = {"AACCGGTA"};
        printf("%d\n", minMutation("AACCGGTT", "AACCGGTA", bank, 1)); // 1
    }
    {
        const char bank[][GENE_LEN + 1] = {"AACCGGTA", "AACCGCTA", "AAACGGTA"};
        printf("%d\n", minMutation("AACCGGTT", "AAACGGTA", bank, 3)); // 2
    }
    {
        const char bank[][GENE_LEN + 1] = {"AAAACCCC", "AAACCCCC", "AACCCCCC"};
        printf("%d\n", minMutation("AAAAACCC", "AACCCCCC", bank, 3)); // 3
    }
    {
        const char bank[][GENE_LEN + 1] = {""};
        printf("%d\n", minMutation("AACCGGTT", "AACCGGTA", bank, 0)); // -1
    }
    return 0;
}
